apis/python/src/tiledbsoma/tiledb_object.py

```diff
@@ -133,6 +143,80 @@ def exists(self) -> bool:
         except tiledb.cc.TileDBError:
             return False
 
+    _open_mode: Optional[str] = None
+    _open_timestamp: Optional[int] = None
+
+    def open(
+        self: TTileDBObject, mode: str = "r", timestamp: Optional[int] = None
+    ) -> TTileDBObject:
+        """
+        Open the object for timestamped read or write operations. The object should be closed when
+        done. It's recommended to use the object as a context manager to automate this, e.g.:
+
+          with tiledbsoma.SparseNDArray(uri=...).open() as X:
+              data = X.read_table(...)
+              data = X.read_table(...)
+
+        ``mode='r'`` (default): opening at a given timestamp (default: now) provides snapshot
+        consistency over a series of read operations on the current object and (for collections)
+        accessed elements.
+
+        ``mode='w'``: opening an object for writing provides that multiple write operations on the
+        object all share the same timestamp. (To write a collection with timestamps shared amongst
+        all elements, explicitly open each element with the desired timestamp.)
+
+        Opening an object for a series of read or write operations can also make them more
+        efficient, by reusing resources and metadata between operations.
+        """
+        if self.is_open():
+            raise RuntimeError("TileDBObject is already open")
+        if mode not in ("r", "w"):
+            raise ValueError("unknown mode")
+        self._open_mode = mode
+        if timestamp is not None:
+            self._open_timestamp = timestamp
+        elif mode == "r" and self._parent_timestamp is not None:
+            # Auto-inherit the parent timestamp only for reads, for now.
+            # TODO: we'd want to inherit the parent timestamp for writes only if the parent itself
+            # is currently open for writing. We don't currently store the reference to parent
+            # beyond the initializer. What if parent is currently open for reading, not writing?
+            # What if parent isn't open but grandparent is? All these complications while the
+            # typical realistic use case is that the array is created and written first, and only
+            # thereafter added to a parent collection. Punting for now...
+            self._open_timestamp = self._parent_timestamp
+        else:
+            self._open_timestamp = int(time.time() * 1000)
+        return self
+
+    def is_open(self) -> bool:
+        return self._open_mode is not None
+
+    def _effective_timestamp(self) -> Optional[int]:
+        if self.is_open():
+            assert self._open_timestamp is not None
+            return self._open_timestamp
+        return self._parent_timestamp
+
+    def close(self) -> None:
+        """
+        Release any resources held while the object is open.
+        """
+        if not self.is_open():
+            raise RuntimeError("TileDBObject is not open")
+        self._open_mode = None
```

Comment:
Should this be cascaded to children?

Reply:
Also, should the underlying tiledb object be closed as well at this point?

Reply:
To avoid some of the edge cases discussed, e.g., client code holding onto TileDbObject references and using them outside of the open() call, I'm wondering if TileDBObjects should raise an error on future reads if the object has been explicitly closed.

Reply:
I am *moderately* in favor of cascaded close operations. My thought there is that given that you’ll mostly be implicitly opening children by accessing them from a parent, closing that parent should also close the children. Provided that we make `close` idempotent, a user closing a child and then closing a parent later would not be a problem.

I am also strongly in favor of saying “once you `close`, you can’t reopen”, because that seems like the general pattern in e.g. built-in `open`.